iceoryx2-bb/cxx/include/iox2/container/static_vector.hpp

```diff
@@ -0,0 +1,319 @@
+            : m_parent(&parent) {
+        }
+
+      public:
+        ~UncheckedAccessor() = default;
+        UncheckedAccessor(UncheckedAccessor const&) = delete;
+        UncheckedAccessor(UncheckedAccessor&&) = delete;
+        auto operator=(UncheckedAccessor const&) -> UncheckedAccessor& = delete;
+        auto operator=(UncheckedAccessor&&) -> UncheckedAccessor& = delete;
+
+        constexpr auto operator[](SizeType index) -> Reference {
+            return *m_parent->m_storage.pointer_from_index(index);
+        }
+
+        constexpr auto operator[](SizeType index) const -> ConstReference {
+            return *m_parent->m_storage.pointer_from_index(index);
+        }
+
+        constexpr auto begin() noexcept -> Iterator {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr auto begin() const noexcept -> ConstIterator {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr auto end() noexcept -> Iterator {
+            return m_parent->m_storage.pointer_from_index(this->m_parent->m_storage.size());
+        }
+
+        constexpr auto end() const noexcept -> ConstIterator {
+            return m_parent->m_storage.pointer_from_index(this->m_parent->m_storage.size());
+        }
+
+        constexpr auto data() noexcept -> Pointer {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr auto data() const noexcept -> ConstPointer {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+    };
+
+  private:
+    template <typename, uint64_t>
+    friend class StaticVector;
+    detail::StaticVectorStorage<T, N> m_storage;
+
+  public:
+    // constructors
+    constexpr StaticVector() noexcept = default;
+    constexpr StaticVector(StaticVector const&) = default;
+    constexpr StaticVector(StaticVector&&) = default;
+
+    template <uint64_t M, std::enable_if_t<(N >= M), bool> = true>
+    // NOLINTNEXTLINE(hicpp-explicit-conversions), conceptually a copy constructor
+    constexpr StaticVector(StaticVector<T, M> const& rhs)
+        : m_storage(rhs.m_storage) {
+    }
+
+    // destructor
+#if __cplusplus >= 202002L
+    constexpr
+#endif
+        ~StaticVector() = default;
+
+    auto operator=(StaticVector const&) -> StaticVector& = delete;
+    auto operator=(StaticVector&&) -> StaticVector& = delete;
+
+    template <typename... Args>
+    constexpr auto try_emplace_back(Args&&... args) ->
+        // NOLINTNEXTLINE(modernize-type-traits), _v requires C++17
+        std::enable_if_t<std::is_constructible<T, Args...>::value, bool> {
+        if (m_storage.size() < N) {
+            new (m_storage.pointer_from_index(m_storage.size)) T(std::forward<Args>(args)...);
+            m_storage.increment_size();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    constexpr auto try_push_back(T const& value) -> bool {
```

Comment:
Food for thought. In iceoryx classic, the philosophy was that it should be easy to port from the STL to our data types but not necessary the other way around. This meant that we had a `push_back` that did not throw exceptions but return a bool. In hindsight, this might have been not the best idea, since there was a semantic change but the code still compiled. The question is if there is some middleground, e.g. having a `push_back(T& v) -> bool` but with a deprecation annotation?

Reply:
I agree that giving the same name to an operation with different semantics is a recipe for trouble. If we aim for consistency here, I guess the closest would be C++26's `inplace_vector`, which provides this operation under the same name, but returns a `pointer` to the inserted element instead of a `bool`. `inplace_vector` still throws `bad_alloc` on `push_back`, precisely for the reasons you mentioned.

I don't understand how a deprecation annotation would solve this problem. You would then get a compiler warning for every call to `push_back`, how is that useful?